File: quick_sort.go
package algorithm

func QuickSort(a []int) {
	if len(a) <= 1 {
		return
	}

	left := 0
	right := len(a) - 1
	x := a[left]

	for left < right {
		for left < right && x < a[right] {
			right--
		}
		if left < right {
			a[left] = a[right]
			left++
		}
		for left < right && x > a[left] {
			left++
		}
		if left < right {
			a[right] = a[left]
			right--
		}
	}
	a[left] = x

	QuickSort(a[:left])
	QuickSort(a[right+1:])
}

File: quick_sort.c

int _partition(int* nums, int left, int right) {
    int x = nums[left]; // TODO switch with a random one
    int tmp;

    // partition
    int p = left, q = right;
    while(p < q) {
        while (p < q && x <= nums[q]) {
            q--;
        }
        nums[p] = nums[q];
        if (p < q) p++;
        while (p < q && nums[p] >= x) {
            p++;
        }
        nums[q] = nums[p];
        if (p < q) q--;
    }

    nums[p] = x;
    return p;
}

void _quick_sort(int* nums, int left, int right) {
    if (left >= right) {
        return;
    }

    int p = _partition(nums, left, right);

    _quick_sort(nums, 0, p-1);
    _quick_sort(nums, p+1, right);
}

void quick_sort(int* nums, int n) {
    _quick_sort(nums, 0, n-1);
}


